aten/src/ATen/native/quantized/cpu/q_adaavgpool.cpp

```diff
@@ -0,0 +1,245 @@
+        }
+      }
+    }
+  });
+}
+
+std::vector<int64_t> get_output_shape(const Tensor& input, IntArrayRef output_size) {
+  for (int64_t i = 0; i < input.dim(); i++) {
+    TORCH_CHECK(
+        input.size(i) > 0,
+        "adaptive_avg_pooling2d(): expected input to have non-empty spatial "
+        "dimensions, but input has sizes ",
+        input.sizes(),
+        " with dimension ",
+        i,
+        " being empty");
+  }
+
+  TORCH_CHECK(
+      (input.dim() == 3 || input.dim() == 4),
+      "non-empty 3D or 4D (batch mode) tensor expected for input");
+
+  /* sizes */
+  int64_t sizeD = input.size(-3);
+  const auto osizeH = output_size[0];
+  const auto osizeW = output_size[1];
+
+  /* resize output */
+  std::vector<int64_t> output_shape;
+  int64_t sizeB = 0;
+  if (input.dim() == 3) {
+    output_shape = {sizeD, osizeH, osizeW};
+  } else {
+    sizeB = input.size(-4);
+    output_shape = {sizeB, sizeD, osizeH, osizeW};
+  }
+
+  return output_shape;
+}
+
+template <typename scalar_t>
+Tensor q_adaptive_avg_pool2d(
+    const Tensor& input,
+    IntArrayRef output_size) {
+  Tensor output;
+  const auto output_shape = get_output_shape(input, output_size);
+  /* sizes */
+  int64_t sizeD = input.size(-3);
+  int64_t isizeH = input.size(-2);
+  int64_t isizeW = input.size(-1);
+  /* strides */
+  int64_t istrideD = input.stride(-3);
+  int64_t istrideH = input.stride(-2);
+  int64_t istrideW = input.stride(-1);
+
+  auto osizeH = output_shape[output_shape.size() - 2];
+  auto osizeW = output_shape[output_shape.size() - 1];
+  int64_t sizeB = output_shape.size() == 3 ? 0 : output_shape[0];
+
+  if (input.is_contiguous(c10::MemoryFormat::ChannelsLast)) {
+    // Fast path for NHWC
+    Tensor output = at::_empty_affine_quantized(
+        output_shape,
+        input.options(),
+        input.q_scale(),
+        input.q_zero_point(),
+        input.suggest_memory_format());
+    if (input.dim() == 3 || input.size(0) == 1) {
+      qadaptive_avg_pool2d_nhwc_stub(
+          input.device().type(),
+          input,
+          output,
+          0,
+          sizeD,
+          isizeH,
+          isizeW,
+          osizeH,
+          osizeW,
+          0,
+          istrideD,
+          istrideH,
+          istrideW);
+    } else {
+      int64_t istrideB = input.stride(-4);
+      at::parallel_for(0, sizeB, 0, [&](int64_t start, int64_t end) {
+        for (auto b = start; b < end; b++) {
+          qadaptive_avg_pool2d_nhwc_stub(
+              input.device().type(),
+              input,
+              output,
+              b,
+              sizeD,
+              isizeH,
+              isizeW,
+              osizeH,
+              osizeW,
+              istrideB,
+              istrideD,
+              istrideH,
+              istrideW);
+        }
+      });
+    }
+    return output;
+  } else {
+    Tensor output = at::_empty_affine_quantized(
```

Comment:
In this branch `input` might be non contiguous, is it intended or you need to call `input = intput.contiguous()` in this case.

Reply:
Nevermind, see next comment instead.